assignment1/src/app/task/task.service.ts

```diff
@@ -0,0 +1,89 @@
+import {EventEmitter, Injectable} from '@angular/core';
+import {ListTasks} from '../shared/task.model';
+import {Subject} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TaskService {
+  newListTask = new Subject<ListTasks[]>();
+  listTask: ListTasks[] = this.onGetListTask();
+  defaultListTask: ListTasks = {
+    categoryName: 'category 1',
+    tasks: [{name: 'task 1', deadline: new Date(), createdAt: new Date(), status: 'available'}]
+  };
+
+  constructor() {
+  }
+
+  onGetListTask(): ListTasks[] {
+    return JSON.parse(localStorage.getItem('database') || '[]');
+  }
+
+  onSetListTask() {
+    localStorage.setItem('database', JSON.stringify(this.listTask));
+  }
+
+  onSaveAndBindingNewListTask() {
+    this.onSetListTask();
+    this.newListTask.next(this.listTask);
+  }
+
+  onChangeStatusTaskById(idListTask: number, idTask: number, newStatus: string) {
+    this.listTask[idListTask].tasks[idTask].status = newStatus;
```

Comment:
this.listTask[idListTask].tasks[idTask].status

Expression like this will easily cause `undefined` error.

Reply:
i'm writting function check for all longer access object, and then use it in if, else check of function actions
`  onCheckPropertyExits(idListTask: number, idTask: number = -1) {
    if (idTask === -1) {
      return this.listTask[idListTask];
    } else {
      return !(!this.listTask[idListTask] || !this.listTask[idListTask].tasks[idTask]);
    }
  }`